iroh-io/src/lib.rs

```diff
@@ -0,0 +1,603 @@
+/// length of the resource.
+///
+/// This is similar to the io interface of sqlite.
+/// See xRead, xFileSize in <https://www.sqlite.org/c3ref/io_methods.html>
+#[allow(clippy::len_without_is_empty)]
+pub trait AsyncSliceReader {
+    /// The future returned by read_at
+    type ReadAtFuture<'a>: Future<Output = io::Result<Bytes>> + 'a
+    where
+        Self: 'a;
+    /// Read the entire buffer at the given position.
+    ///
+    /// Will return at most `len` bytes, but may return fewer if the resource is smaller.
+    /// If the range is completely covered by the resource, will return exactly `len` bytes.
+    /// If the range is not covered at all by the resource, will return an empty buffer.
+    /// It will never return an io error independent of the range as long as the underlying
+    /// resource is valid.
+    #[must_use = "io futures must be polled to completion"]
+    fn read_at(&mut self, offset: u64, len: usize) -> Self::ReadAtFuture<'_>;
+
+    /// The future returned by len
+    type LenFuture<'a>: Future<Output = io::Result<u64>> + 'a
+    where
+        Self: 'a;
+    /// Get the length of the resource
+    #[must_use = "io futures must be polled to completion"]
+    fn len(&mut self) -> Self::LenFuture<'_>;
+}
+
+/// A trait to abstract async writing to different resources.
+///
+/// This trait does not require the notion of a current position, but instead
+/// requires explicitly passing the offset to write_at and write_array_at.
+/// In addition to the ability to write at an arbitrary offset, it also provides
+/// the ability to set the length of the resource.
+///
+/// This is similar to the io interface of sqlite.
+/// See xWrite in <https://www.sqlite.org/c3ref/io_methods.html>
+pub trait AsyncSliceWriter: Sized {
+    /// The future returned by write_at
+    type WriteAtFuture<'a>: Future<Output = io::Result<()>> + 'a
+    where
+        Self: 'a;
+    /// Write the entire Bytes at the given position.
+    ///
+    /// When writing with the end of the range after the end of the blob, the blob will be extended.
+    /// When writing with the start of the range after the end of the blob, the gap will be filled with zeros.
+    #[must_use = "io futures must be polled to completion"]
+    fn write_at(&mut self, offset: u64, data: Bytes) -> Self::WriteAtFuture<'_>;
+
+    /// The future returned by write_slice_at
+    type WriteArrayAtFuture<'a>: Future<Output = io::Result<()>> + 'a
+    where
+        Self: 'a;
+    /// Write the entire fixed size array at the given position.
+    /// This is useful to write a small number of bytes without allocating a [bytes::Bytes].
+    ///
+    /// Except for taking a fixed size array instead of a [bytes::Bytes], this is equivalent to [AsyncSliceWriter::write_at].
+    #[must_use = "io futures must be polled to completion"]
+    fn write_array_at<const N: usize>(
+        &mut self,
+        offset: u64,
+        data: [u8; N],
+    ) -> Self::WriteArrayAtFuture<'_>;
```

Comment:
Could be nice to provide a default impl that just fallsback to use `write_at`

Reply:
😢 default impls would require defaults for associated types, which are unstable

Reply:
😭